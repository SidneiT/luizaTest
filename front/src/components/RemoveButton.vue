<template>
  <button v-on="$listeners" type="button" class="remove-button" title="Lista de Desejos">
    <span :class="['material-icons', {'red': selected}]">
      highlight_off
    </span>
  </button>
</template>

<script>
export default {
  name: 'RemoveButton',
  props: {
    selected: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<style lang="scss" scoped>
.remove-button {
  position: absolute;
  top: 4px;
  right: 3px;
  color: black;
  border: none;
  background-color: transparent;
  outline: none;

  span.material-icons {
    font-size: 34px;
  }

  &:hover {
    cursor: pointer;
  }
}
</style>
