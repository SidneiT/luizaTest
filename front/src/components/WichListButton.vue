<template>
  <button v-on="$listeners" type="button" class="product-wichlist-button" title="Lista de Desejos">
    <span :class="['material-icons', {'red': selected}]">
      favorite
    </span>
  </button>
</template>

<script>
export default {
  name: 'WichiListButton',
  props: {
    selected: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<style lang="scss" scoped>
.product-wichlist-button {
  position: absolute;
  top: -3px;
  left: 4px;
  background-color: black;
  border: none;
  height: 40px;
  width: 40px;
  border-radius: 3px;

  &:hover {
    cursor: pointer;
  }

  span{
    transition: color 0.2s linear;

    &.material-icons{
      color: white;
      margin-top: 10px;
    }

    &.red {
      color: #d52222;
    }
  }

  &::after {
    position: inherit;
    left: 2px;
    top: 40px;
    content: "";
    border-left: 18px solid transparent;
    border-right: 18px solid transparent;
    border-top: 10px solid black;
    border-bottom-left-radius: 50%;
  }
}
</style>
