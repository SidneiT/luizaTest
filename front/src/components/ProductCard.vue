<template>
  <div class="product-card">
    <img :src="img" :alt="title" />
    <span class="product-card__title">{{title}}</span>
    <span class="product-card__price">{{priceFormated}}</span>
    <WichListButton v-if="showWichListButton" :selected="wichList" @click="toggleWichList"/>
    <RemoveButton v-if="showRemoveButton" @click="removeFromWichList(productId)"/>
  </div>
</template>

<script>
import WichListButton from '../components/WichListButton.vue'
import RemoveButton from '../components/RemoveButton.vue'
import { mapActions } from 'vuex'

export default {
  name: 'ProductCard',
  components: {
    WichListButton,
    RemoveButton
  },
  props: {
    img: {
      type: String,
      required: true
    },
    title: {
      type: String,
      required: true
    },
    productId: {
      type: Number,
      required: true
    },
    price: {
      type: Number,
      required: true
    },
    currency: {
      type: String,
      required: true
    },
    wichList: {
      type: Boolean,
      default: undefined
    },
    wichListOptions: {
      type: Array,
      validator: values => {
        const options = ['flag-favorite', 'circle-remove']

        const invalid = values.filter(v => !options.includes(v))

        return !invalid.length
      }
    }
  },
  computed: {
    priceFormated () {
      return this.price.toLocaleString(this.currency, {
        style: 'currency',
        currency: this.currency
      })
    },
    showWichListButton () {
      return this.wichListOptions.includes('flag-favorite')
    },
    showRemoveButton () {
      return this.wichListOptions.includes('circle-remove')
    }
  },
  methods: {
    ...mapActions(['addToWichList', 'removeFromWichList']),
    toggleWichList () {
      this.wichList
        ? this.removeFromWichList(this.productId)
        : this.addToWichList(this.productId)
    }
  }
}
</script>

<style lang="scss" scoped>
.product-card {
  position: relative;
  width: 200px;
  height: 350px;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  border: 1px solid #dcdcdc;
  border-radius: 5px;

  img {
    max-width: 100%;
    border-top-left-radius: 5px;
    border-top-right-radius: 5px;
  }

  &__title {
    font-weight: bold;
    padding: 15px;
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
    width:calc(100% - 30px);
  }

  &__price {
    display: flex;
    align-items: center;
    flex: 1 1 auto;
    color: #f4d957;
    font-weight: bold;
  }
}
</style>
