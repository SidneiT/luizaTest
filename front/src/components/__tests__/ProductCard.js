/* eslint-disable no-debugger */
import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'
import ProductCard from '../../components/ProductCard.vue'
import WichListButton from '../../components/WichListButton.vue'
import RemoveButton from '../../components/RemoveButton.vue'

const localVue = createLocalVue()
localVue.use(Vuex)

const propsData = {
  productId: 0,
  title: 'Camisa Nike Corinthians I',
  price: 229.9,
  currency: 'BRL',
  img: 'https://via.placeholder.com/300x300',
  wichList: true,
  wichListOptions: ['flag-favorite']
}

describe('ProductCard.vue', () => {
  let store
  let actions

  beforeEach(() => {
    actions = {
      removeFromWichList: jest.fn(() => {})
    }

    store = new Vuex.Store({ actions })
  })

  test('renders correctly', () => {
    const wrapper = shallowMount(ProductCard, { propsData })

    expect(wrapper.element).toMatchSnapshot()
  })

  test('contains wichlist button', () => {
    const wrapper = shallowMount(ProductCard, { propsData })

    const btn = wrapper.findComponent(WichListButton)
    const btnRemove = wrapper.findComponent(RemoveButton)

    expect(btn.exists()).toBe(true)
    expect(btnRemove.exists()).toBe(false)
  })

  test('Button remove call remove action on click', async () => {
    propsData.wichListOptions = ['circle-remove']

    const wrapper = shallowMount(ProductCard, {
      propsData,
      localVue,
      store
    })

    const btn = wrapper.findComponent(RemoveButton)

    expect(btn.exists()).toBe(true)

    btn.vm.$emit('click')

    expect(actions.removeFromWichList).toHaveBeenCalled()
  })
})
