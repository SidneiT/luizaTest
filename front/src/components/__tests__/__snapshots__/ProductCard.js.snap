// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard.vue renders correctly 1`] = `
<div
  class="product-card"
>
  <img
    alt="Camisa Nike Corinthians I"
    src="https://via.placeholder.com/300x300"
  />
   
  <span
    class="product-card__title"
  >
    Camisa Nike Corinthians I
  </span>
   
  <span
    class="product-card__price"
  >
    R$Â 229,90
  </span>
   
  <wichlistbutton-stub
    selected="true"
  />
   
  <!---->
</div>
`;
